### PR TITLE
[feat]: [CDS-82763]: Added validations and update green service flow.

### PR DESCRIPTION
**Pull request by PragyeshMishra01**
## Describe your changes
1. Added update green service flow. Currently in ECS Blue Green Deployments, we delete the existing green service and we create new green service with same name. Now, we are providing an checkbox in UI for update green service. If this checkbox is enabled, we will not delete existing green service and will update this service with new artifact and configurations.
2. Added Blue Green Service validations. Basically, we add tags to ECS Service in Blue Green Deployments and based on this tag, we determine whether service is blue or green. At the end of deployment, we swap the tags between blue and green service. So, we have observed that if user aborts an ongoing deployment through harness and tags maybe messed up, then in subsequent deployments, we can't predict blue/green service accurately. So, we have added some validations before starting deployment. we will validate if tags are correct or not on basis of load balancer configurations. We will fix the tag if possible otherwise will fail the execution incase we found  load balancer invalid configurations.
3. Introduced a FF for above validations as we wanted to be on safe side and as this is a change in behaviour.
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55783)

### CODE REVIEW

**Comment by PragyeshMishra01** [comment]
trigger uts

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java [hunk 1]
```diff
@@ -1274,6 +1325,91 @@ private boolean checkForDefaultRule(
...
+    if (firstService.isPresent()) {
+      firstServiceTargetGroups = validateTargetGroup(infraConfig, loadBalancerConfig, firstService.get(), logCallback);
+    }
+
+    if (secondService.isPresent()) {
+      secondServiceTargetGroups =
+          validateTargetGroup(infraConfig, loadBalancerConfig, secondService.get(), logCallback);
+    }
+
+    if (CollectionUtils.isEmpty(firstServiceTargetGroups) || CollectionUtils.isEmpty(secondServiceTargetGroups)) {
```
Comment (ivan-83) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java:1350]: Why empty if statement?

File: 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java [hunk 2]
```diff
@@ -937,8 +937,10 @@ public enum FeatureName {
   CDS_APPROVAL_AND_STAGE_NOTIFICATIONS_WITH_CD_METADATA(
       "Improves approval and stage notifications with cd metadata", HarnessTeam.CDC),
   GTM_DEVELOPER_LICENSING("Enable new developer licensing changes", GTM),
-  CDS_ASG_SHIFT_TRAFFIC_STEP_NG("Enable Shift Traffic step for Asg Blue-Green deployments", CDC);
-
+  CDS_ASG_SHIFT_TRAFFIC_STEP_NG("Enable Shift Traffic step for Asg Blue-Green deployments", CDC),
+  CDS_ECS_BG_VALIDATION("Enable some conditions, which we will execute before the deployment. "
```
Comment (ivan-83) [inline 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java:941]: Do we have approval for FF? Can we be more specific in the description of which condition?

**Comment by ivan-83** [comment]
Can we add a description on PR, what is "update green service flow" and why are we doing validation?

**Comment by PragyeshMishra01** [comment]
trigger uts

File: 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java (hunk 2 above)
Comment (PragyeshMishra01) [reply 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java:943]: yes, we have approval from @thisrohangupta 
REF: https://harness.atlassian.net/browse/CDS-81314?focusedCommentId=613674

Comment (PragyeshMishra01) [reply 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java:943]: also added detailed description.

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java (hunk 1 above)
Comment (PragyeshMishra01) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java:1350]: added **return;** in if statement  for better readibility. 
In this if, we are ignoring further logic incase target group is null.

**Comment by PragyeshMishra01** [comment]
trigger uts

**Comment by PragyeshMishra01** [comment]
trigger uts

**Comment by PragyeshMishra01** [comment]
trigger uts

**Review by piyush-bhuwalka — APPROVED** [review]

**Review by ivan-83 — CHANGES_REQUESTED** [review]
> Reviewed 8 of 10 files at r1, 1 of 2 files at r2, 2 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55783)* status: 13 file(s) to review. 1 change requests. 1 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java` line 1333 at r4](https://reviewable.io/reviews/harness/harness-core/55783#-NjxVNau5p4t4mBNOq5k:-NjxVNau5p4t4mBNOq5l:bykqey5) ([raw file](https://github.com/harness/harness-core/blob/93e58aa3b77f72701a1ff7720981b485d9cca0d3/930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java#L1333)):*
> ```Java
>     String firstVersionServiceName = servicePrefix + 1;
>     String secondVersionServiceName = servicePrefix + 2;
>     Optional<Service> firstService = describeService(
> ```

Can we rename the first and second services based on versions? 
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java` line 1353 at r4](https://reviewable.io/reviews/harness/harness-core/55783#-NjxWCDY3_iSoJ6JBpQ7:-NjxWCDY3_iSoJ6JBpQ8:b-2ixlad) ([raw file](https://github.com/harness/harness-core/blob/93e58aa3b77f72701a1ff7720981b485d9cca0d3/930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java#L1353)):*
> ```Java
>       return;
>     } else if (loadBalancerConfig.getProdTargetGroupArn().equals(loadBalancerConfig.getStageTargetGroupArn())) {
>       if (firstService.get().desiredCount() > secondService.get().desiredCount()) {
> ```

Can we separate logic for validation and for updating in separate methods?
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java` line 1368 at r4](https://reviewable.io/reviews/harness/harness-core/55783#-NjxUa9O8RofVVmHvjP8:-NjxUa9P73CWeEGBn4ih:b2041of) ([raw file](https://github.com/harness/harness-core/blob/93e58aa3b77f72701a1ff7720981b485d9cca0d3/930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java#L1368)):*
> ```Java
>                 secondService.get().serviceName(), loadBalancerConfig.getProdTargetGroupArn(), PROD_SERVICE,
>                 PROD_SERVICE, PROD_SERVICE),
>             new InvalidRequestException(INVALID_LOAD_BALANCER_CONFIG_ERROR));
> ```

This could be a separate method that accepts loadBalancerConfig.getProdTargetGroupArn() and it should work for the below "else if" to avoid code duplication.



_Code quote:_
```Java
      if (firstServiceTargetGroups.contains(loadBalancerConfig.getProdTargetGroupArn())
          && secondServiceTargetGroups.contains(loadBalancerConfig.getProdTargetGroupArn())) {
        throw NestedExceptionUtils.hintWithExplanationException(
            format(BOTH_SERVICE_WITH_SAME_TARGET_GROUP_ERROR_HINT, PROD_SERVICE),
            format(BOTH_SERVICE_WITH_SAME_TARGET_GROUP_ERROR_EXPLANATION, firstService.get().serviceName(),
                secondService.get().serviceName(), loadBalancerConfig.getProdTargetGroupArn(), PROD_SERVICE,
                PROD_SERVICE, PROD_SERVICE),
            new InvalidRequestException(INVALID_LOAD_BALANCER_CONFIG_ERROR));
```
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java` line 1381 at r4](https://reviewable.io/reviews/harness/harness-core/55783#-NjxWWKa-OX05Xz99DuN:-NjxWWKa-OX05Xz99DuO:b3p5mx) ([raw file](https://github.com/harness/harness-core/blob/93e58aa3b77f72701a1ff7720981b485d9cca0d3/930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java#L1381)):*
> ```Java
>   }
> 
>   private List<String> validateTargetGroup(
> ```

This method does validate and update. 
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java` line 1400 at r4](https://reviewable.io/reviews/harness/harness-core/55783#-NjxWo5_35fW9sHmtCXj:-NjxWo5_35fW9sHmtCXk:b-9iy23y) ([raw file](https://github.com/harness/harness-core/blob/93e58aa3b77f72701a1ff7720981b485d9cca0d3/930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java#L1400)):*
> ```Java
>               loadBalancerConfig.getStageTargetGroupArn()),
>           new InvalidRequestException(INVALID_LOAD_BALANCER_CONFIG_ERROR));
>     } else if (targetGroups.contains(loadBalancerConfig.getProdTargetGroupArn())) {
> ```

can we separate the logic for validation and update?

**Comment by PragyeshMishra01** [comment]
trigger uts

**Review by ivan-83 — APPROVED** [review]
> Reviewed 2 of 2 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55783)* status: 13 file(s) to review. 1 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java` line 1356 at r6](https://reviewable.io/reviews/harness/harness-core/55783#-NkGB624BoFWA2P23ef0:-NkGB624BoFWA2P23ef1:boghj7j) ([raw file](https://github.com/harness/harness-core/blob/1be0be987a67b69ccd52c9293ca6edd6a4a0cf19/930-delegate-tasks/src/main/java/io/harness/delegate/task/ecs/EcsCommandTaskNGHelper.java#L1356)):*
> ```Java
>             loadBalancerConfig.getProdTargetGroupArn());
>       } else if (firstServiceTargetGroups.contains(loadBalancerConfig.getStageTargetGroupArn())
>           && secondServiceTargetGroups.contains(loadBalancerConfig.getStageTargetGroupArn())) {
> ```

firstServiceTargetGroups.contains(loadBalancerConfig.getStageTargetGroupArn())
          && secondServiceTargetGroups.contains(loadBalancerConfig.getStageTargetGroupArn())

Minor: could be part of first if as or(||)

**Review by piyush-bhuwalka — APPROVED** [review]

**Comment by PragyeshMishra01** [comment]
trigger review

**Comment by PragyeshMishra01** [comment]
trigger ut3

**Pull request merged**